smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,235 @@
+            assert discoveredWebsocketEndpoints == null;
+
+            List<SmackFuture<LookupConnectionEndpointsResult, Exception>> futures = new ArrayList<>(2);
+
+            InternalSmackFuture<LookupConnectionEndpointsResult, Exception> websocketEndpointsLookupFuture = new InternalSmackFuture<>();
+
+            connectionInternal.asyncGo(() -> {
+
+                // Look for connection endpoints, the result obtained will contain the endpoints obtained in the lookup process.
+                Result<WebsocketRemoteConnectionEndpoint> result = RemoteWebsocketConnectionEndpoints.lookup(connectionInternal);
+
+                LookupConnectionEndpointsResult endpointsResult;
+
+                // If discoveredRemoteConnectionEndpoints list is empty set the result to WebsocketEndpointsDiscoveryFailed,
+                // or otherwise to DiscoveredWebsocketEndpoints.
+                if (result.discoveredRemoteConnectionEndpoints.isEmpty()) {
+                    endpointsResult = new WebsocketEndpointsDiscoveryFailed(result);
+                } else {
+                    endpointsResult = new DiscoveredWebsocketEndpoints(result);
+                }
+                websocketEndpointsLookupFuture.setResult(endpointsResult);
+            });
+
+            futures.add(websocketEndpointsLookupFuture);
+            return futures;
+        }
+
+        @Override
+        protected void loadConnectionEndpoints(LookupConnectionEndpointsSuccess lookupConnectionEndpointsSuccess) {
+            discoveredWebsocketEndpoints = (DiscoveredWebsocketEndpoints) lookupConnectionEndpointsSuccess;
+        }
+
+        @Override
+        protected void afterFiltersClosed() {
+        }
+
+        @Override
+        protected void disconnect() {
+            connectionAttemptState.websocket.disconnect();
+        }
+
+        @Override
+        protected void notifyAboutNewOutgoingElements() {
+            afterOutgoingElementsQueueModified();
+        }
+
+        @Override
+        public SSLSession getSslSession() {
+            return null;
+        }
+
+        @Override
+        public boolean isConnected() {
+            return connectionInternal.connection.isConnected();
+        }
+
+        @Override
+        public Stats getStats() {
+            return null;
+        }
+
+        public final class DiscoveredWebsocketEndpoints implements LookupConnectionEndpointsSuccess {
+            final RemoteWebsocketConnectionEndpoints.Result<WebsocketRemoteConnectionEndpoint> result;
+
+            DiscoveredWebsocketEndpoints(Result<WebsocketRemoteConnectionEndpoint> result) {
+                this.result = result;
+            }
+
+            public RemoteWebsocketConnectionEndpoints.Result<WebsocketRemoteConnectionEndpoint> getResult() {
+                return result;
+            }
+        }
+
+        final class WebsocketEndpointsDiscoveryFailed implements LookupConnectionEndpointsFailed {
```

Comment:
You probably want to sensibly overwrite the `toString()` method of this class, as right now connection failures yield this not so helpful error message:

```
Exception in thread "main" org.jivesoftware.smack.SmackException$NoEndpointsDiscoveredException: Not endpoints could be discovered due the following lookup failures: org.jivesoftware.smack.websocket.XmppWebsocketTransportModule$XmppWebsocketTransport$WebsocketEndpointsDiscoveryFailed@65c7a252
```
Information about the lookup failures would be desirable.

Also I noticed that @Flowdalic left a TODO in [the LookupConnectionEndpointsFailed](https://github.com/igniterealtime/Smack/blob/master/smack-core/src/main/java/org/jivesoftware/smack/c2s/XmppClientToServerTransport.java#L73) class, which make me wonder if you could maybe come up with some methods that interface may want to declare.

Reply:
I have overridden `toString()` here and now we can see the cause for the failure.
I also went through the `TODO` inside `LookupConnectionEndpointFailed` and I feel that adding a method to the interface is probably an overkill for fixing this issue.
Thanks for pointing this out. Let me know if you have any other suggestions :)